src/Microsoft.ML.Featurizers/CategoryImputer.cs

```diff
@@ -0,0 +1,766 @@
+            private static extern bool DestroyEstimatorNative(IntPtr estimator, out IntPtr errorHandle); // Should ONLY be called by safe handle
+
+            [DllImport("Featurizers", EntryPoint = "CatImputerFeaturizer_string_CreateTransformerFromEstimator", CallingConvention = CallingConvention.Cdecl), SuppressUnmanagedCodeSecurity]
+            private static extern bool CreateTransformerFromEstimatorNative(TransformerEstimatorSafeHandle estimator, out IntPtr transformer, out IntPtr errorHandle);
+            internal override void CreateTransformerFromEstimator(IDataView input)
+            {
+                _transformerHandler = CreateTransformerFromEstimatorBase(input);
+            }
+
+            [DllImport("Featurizers", EntryPoint = "CatImputerFeaturizer_string_CreateTransformerFromSavedData", CallingConvention = CallingConvention.Cdecl), SuppressUnmanagedCodeSecurity]
+            private static unsafe extern bool CreateTransformerFromSavedDataNative(byte* rawData, IntPtr bufferSize, out IntPtr transformer, out IntPtr errorHandle);
+            private protected override unsafe void CreateTransformerFromSavedDataHelper(byte* rawData, IntPtr dataSize)
+            {
+                var result = CreateTransformerFromSavedDataNative(rawData, dataSize, out IntPtr transformer, out IntPtr errorHandle);
+                if (!result)
+                    throw new Exception(GetErrorDetailsAndFreeNativeMemory(errorHandle));
+
+                _transformerHandler = new TransformerEstimatorSafeHandle(transformer, DestroyTransformerNative);
+            }
+
+            [DllImport("Featurizers", EntryPoint = "CatImputerFeaturizer_string_DestroyTransformer", CallingConvention = CallingConvention.Cdecl), SuppressUnmanagedCodeSecurity]
+            private static extern bool DestroyTransformerNative(IntPtr transformer, out IntPtr errorHandle);
+
+            [DllImport("Featurizers", EntryPoint = "CatImputerFeaturizer_string_Transform", CallingConvention = CallingConvention.Cdecl), SuppressUnmanagedCodeSecurity]
+            private static unsafe extern bool TransformDataNative(TransformerEstimatorSafeHandle transformer, byte* input, out IntPtr output, out IntPtr outputSize, out IntPtr errorHandle);
+            [DllImport("Featurizers", EntryPoint = "CatImputerFeaturizer_string_DestroyTransformedData", CallingConvention = CallingConvention.Cdecl), SuppressUnmanagedCodeSecurity]
+            private static extern bool DestroyTransformedDataNative(IntPtr output, IntPtr outputSize, out IntPtr errorHandle);
+            internal unsafe override ReadOnlyMemory<char> Transform(ReadOnlyMemory<char> input)
+            {
+                var inputAsString = input.ToString();
+                fixed (byte* interopInput = string.IsNullOrEmpty(inputAsString) ? null : Encoding.UTF8.GetBytes(inputAsString + char.MinValue))
```

Comment:
Is `Transform()` called for each row to apply the, already learned, imputer? If so, isn't this quite inefficient to reallocate the string for each row of data?

Shouldn't this be roughly:
```C#
return ((!input.IsEmpty || input.Length == 0) ? ReplacementValue : input);
```

Reply:
I need the input string as a byte array so that I can pass it to the native code. Currently, we are not getting the `Replacement` value back from the native code, though that is an interesting optimization that we could potentially look at in the future.

Reply:
I think we're relying on the ONNX code a bit much in that case. This is a lot of setup code to do a really small amount of work (the one line above). This seems to be a bit of the tail-wagging-the-dog style, where the `ML.NET` dependency, ONNX, is driving the `ML.NET` code instead of the other way around. If we need ONNX to take  `ReadOnlyMemory<char>`, it should (and without allocating any new memory in its code). Specifically, there should be no allocations in this function, or the ONNX portion.

We can mitigate the overhead:
The replacement value is static once the transform is trained, and it's trained before hitting this section of the code. 

How about we just cache the value after calling the native ONNX code _once_? 

```C#
// Early out if (1) if a replacement isn't needed or (2) the cached value is known.
// Otherwise continue with existing code and cache the inputation value at the end.
if (!(input.IsEmpty || input.Length == 0))
  return input; // No need to impute
else {
  if (ReplacementValue != null) // Check if the cached replacement value is known
    return ReplacementValue;
}

// Fall through to continue with the existing function
//...

ReplacementValue = valueFromOnnx;
```

Where `ReplacementValue`  is thread local to the transform (to avoid using a lock for race condition in initializing the value; though all threads will fill it with the same value); or more simply just store the value after the transform is trained or initialized.

Reply:
I can easily fix it so that if its not null it just returns the input. I will add that in shortly. The caching of the replacement value though I am not sure how easy/hard that would be to update as the native code itself doens't currently have a way to return that to me. Is that something you are willing to punt on if I fix it so it only calls the native code if there isn't a value in the input?

Reply:
I believe the native code returns the same value each time (for each row with a missing value). Hence all empty/null inputs, for a single feature column, gets filled with the same static value. 

I believe we just have to cache this one value after the 1st (and only) call to the native code -- call it with the 1st missing value, and store the result. I expect there's no changes to the native code.

Reply:
Made the changes you suggest. Returning the input value if its not null/empty, and caching the result after the first call to transform.

Reply:
Thanks for making the change! 

Your new function looks fast and svelte:
https://github.com/dotnet/machinelearning/blob/795559d951366dd126b6d7b092342155459c25d8/src/Microsoft.ML.Featurizers/CategoricalImputer.cs#L686-L692